rts/motoko-rts/src/idl.rs

```diff
@@ -831,20 +1207,25 @@ unsafe extern "C" fn idl_sub_buf_init(rel_buf: *mut u32, typtbl_size1: u32, typt
     rel.init();
 }
 
-#[no_mangle]
-unsafe extern "C" fn idl_sub(
+#[ic_mem_fn]
+unsafe fn idl_sub<M: Memory>(
+    mem: &mut M,
     rel_buf: *mut u32, // a buffer with at least 2 * typtbl_size1 * typtbl_size2 bits
     typtbl1: *mut *mut u8,
-    typtbl2: *mut *mut u8,
     typtbl_end1: *mut u8,
-    typtbl_end2: *mut u8,
     typtbl_size1: u32,
-    typtbl_size2: u32,
+    candid_data2: Value,
+    type_offsets2: Value,
     t1: i32,
     t2: i32,
 ) -> bool {
     debug_assert!(rel_buf != (0 as *mut u32));
 
+    let mut type_descriptor2 = TypeDescriptor::new(candid_data2, type_offsets2);
+    let typtbl2 = type_descriptor2.build_type_table(mem);
```

Comment:
Hmm, I think using object pooling for the global type table still doesn't quite solve the allocation problem.
`build_type_table` allocates a transient blob on each call to idl_sub. 

I think the way to fix this maybe to revert the implementation of idl_sub (to have symmetric arguments as before) and modify deserialization to use three more Registers to reference the global type table, initialized once with all the other registers. The GC won't run during deserialiazation so it should be safe to build_type_table just once for the use off all nested calls to idl_sub.
Have to try this if you don't want to.

Reply:
Thanks a lot for your PR that improved this. I merged it.